src/main/java/gregtech/common/covers/filter/FluidFilterContainer.java

```diff
@@ -1,128 +1,181 @@
+                                .singletonSlotGroup(101))
+                        .onUpdateListener(w -> {
+                            if (!hasFilter() && panel.isPanelOpen()) {
+                                panel.closePanel();
+                            }
+                        }, true)
+                        .size(18).marginRight(2)
+                        .background(GTGuiTextures.SLOT, GTGuiTextures.FILTER_SLOT_OVERLAY))
+                .child(filterButton
+                        .setEnabledIf(w -> hasFilter())
+                        .onMousePressed(i -> {
+                            boolean success = false;
+                            if (!panel.isPanelOpen()) {
+                                panel.openPanel();
+                                success = true;
+                            } else if (panel.isValid()) {
+                                panel.closePanel();
+                                success = true;
+                            }
+                            Interactable.playButtonClickSound();
+                            return success;
+                        }))
+                .child(IKey.dynamic(() -> hasFilter() ?
+                        getFilterInventory().getStackInSlot(0).getDisplayName() :
+                        IKey.lang("metaitem.fluid_filter.name").get())
+                        .alignment(Alignment.CenterRight).asWidget()
+                        .left(36).right(0).height(18));
     }
 
-    public boolean testFluidStack(FluidStack fluidStack) {
-        return filterWrapper.testFluidStack(fluidStack);
+    public void setBucketOnly(boolean bucketOnly) {
+        if (hasFilter()) {
+            getFilter().setBucketOnly(bucketOnly);
+        }
     }
 
-    public boolean testFluidStack(FluidStack fluidStack, Boolean whitelist) {
-        return filterWrapper.testFluidStack(fluidStack, whitelist);
+    @Override
+    public void readInitialSyncData(@NotNull PacketBuffer packetBuffer) {
+        var stack = ItemStack.EMPTY;
+        try {
+            stack = packetBuffer.readItemStack();
+        } catch (IOException ignore) {}
+        this.filterInventory.setStackInSlot(0, stack);
+
+        if (FilterTypeRegistry.isFluidFilter(stack))
+            setFilter(FilterTypeRegistry.getFluidFilterForStack(stack));
     }
 
-    @Override
-    public NBTTagCompound serializeNBT() {
-        NBTTagCompound tagCompound = new NBTTagCompound();
-        tagCompound.setTag("FilterInventory", filterInventory.serializeNBT());
-        tagCompound.setBoolean("IsBlacklist", filterWrapper.isBlacklistFilter());
-        if (filterWrapper.getFluidFilter() != null) {
-            NBTTagCompound filterInventory = new NBTTagCompound();
-            filterWrapper.getFluidFilter().writeToNBT(filterInventory);
-            tagCompound.setTag("Filter", filterInventory);
+    protected void onFilterSlotChange(boolean notify) {
+        ItemStack filterStack = filterInventory.getStackInSlot(0);
```

Comment:
I don't think this is correct. The Filter Stack is never serialized in `FilterInventory`, just the contents of the filter.

Reply:
the serialization of the filter stack is done in `BaseFilterContainer`, which is shared by item/fluid filter containers